src/rust/src/wallet.rs

```diff
@@ -130,38 +159,182 @@ pub enum WalletError {
+                    {
+                        Some(*txid)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            self.spent_notes.retain(|_, v| to_retain.contains(&v.txid));
+            self.nullifiers.retain(|_, v| to_retain.contains(&v.txid));
+            for (txid, n) in self.wallet_received_notes.iter_mut() {
+                // Erase block height information for any received notes
+                // that have been un-mined by the rewind.
+                if !to_retain.contains(txid) {
+                    n.tx_height = None;
+                }
+            }
+            self.last_observed = Some(last_observed);
+
+            Ok(blocks_to_rewind)
+        } else {
+            Err(RewindError::InsufficientCheckpoints(0))
+        }
     }
 
     /// Add note data for those notes that are decryptable with one of this wallet's
-    /// incoming viewing keys to the wallet, and return the indices of the actions
-    /// that we were able to decrypt.
+    /// incoming viewing keys to the wallet, and return the indices of the actions that we
+    /// were able to decrypt.
     pub fn add_notes_from_bundle(
         &mut self,
         txid: &TxId,
         bundle: &Bundle<Authorized, Amount>,
     ) -> Vec<usize> {
         let mut tx_notes = TxNotes {
+            tx_height: None,
             decrypted_notes: BTreeMap::new(),
         };
 
+        // Check for spends of our notes by matching against the nullifiers
+        // we're tracking, and when we detect one, associate the current
+        // txid and action as spending the note.
+        for (action_idx, action) in bundle.actions().iter().enumerate() {
+            let nf = action.nullifier();
+            // find the outpoint where the note being spent in this action was created
+            if let Some(outpoint) = self.nullifiers.get(nf) {
+                let inpoint = InPoint {
+                    txid: *txid,
+                    action_idx,
+                };
+
+                // If we already have an entry for this nullifier that does not correspond
+                // to the current inpoint, then add the existing inpoint and the current
+                // inpoint to the conflicts map. (The existing inpoint will already be
+                // present if this is not the first detected conflict for this nullifier.)
+                if let Some(cur_inpoint) = self.spent_notes.get(outpoint) {
+                    if cur_inpoint != &inpoint {
+                        let conflicts = self.conflicts.entry(*nf).or_insert_with(BTreeSet::new);
+                        conflicts.insert(*cur_inpoint);
+                        conflicts.insert(inpoint);
+                    }
+                }
+
+                // optimistically consider the latest inpoint to be the correct one
+                self.spent_notes.insert(*outpoint, inpoint);
```

Comment:
Scenario: 
* The local wallet creates a transaction spending some notes and adds it to its own mempool. Another transaction spending those notes will not be allowed into the local mempool. The transaction's bundle is added to the Orchard wallet and recorded in the spent_notes map.
* A block arrives containing another transaction spending those notes. This will cause our transaction to be evicted from our local mempool. The transaction's bundle is added to the Orchard wallet, a conflict is recorded, and the spent_notes map is updated with the new transaction being recorded as the spender.
* A reorg occurs that un-mines the block containing the second transaction, which will then be re-added to the local mempool. That transaction's spentness is removed from the spent_notes map. There are now two transactions in the local wallet that will spend the single source note, but no spentness information is recorded. 
* The user makes another transaction spending those notes. That transaction will not be able to enter the local mempool.